src/cpp/shared_core/FileLogDestination.cpp

```diff
@@ -208,18 +208,36 @@ void FileLogDestination::writeLog(LogLevel in_logLevel, const std::string& in_me
       return;
 
    // Lock the mutex before attempting to write.
-   boost::unique_lock<boost::mutex> lock(m_impl->Mutex);
+   try
+   {
+      boost::lock_guard<boost::mutex> lock(m_impl->Mutex);
 
-   // Open the log file if it's not open. If it fails to open, log nothing.
-   if (!m_impl->LogOutputStream && !m_impl->openLogFile())
-      return;
+      // Open the log file if it's not open. If it fails to open, log nothing.
+      if (!m_impl->openLogFile())
+         return;
 
-   // Rotate the log file if necessary. If it fails to rotate, log nothing.
-   if (!m_impl->rotateLogFile())
-      return;
+      // Rotate the log file if necessary. If it fails to rotate, log nothing.
+      if (!m_impl->rotateLogFile())
+         return;
 
-   (*m_impl->LogOutputStream) << in_message;
-   m_impl->LogOutputStream->flush();
+      (*m_impl->LogOutputStream) << in_message;
+      m_impl->LogOutputStream->flush();
+
+      // If the output stream has bad state after writing, it might have been closed. Try re-opening it and writing the
+      // message again. Often it is not possible to tell that a stream has failed until a write is attempted.
+      if (!m_impl->LogOutputStream->good())
+      {
+         if (!m_impl->openLogFile())
+            return;
+
+         (*m_impl->LogOutputStream) << in_message;
+         m_impl->LogOutputStream->flush();
```

Comment:
Would it be worth checking if we get another log failure here? Is there anything we can or should do if both attempts to write to the logfile fail?

Reply:
If both attempts fail, I don't think a third attempt to open it will work, and we can't attempt to log the failure because that will result in recursive logging attempts, even if there is another log destination available. 

These changes are primarily to address what happens when we daemonize. We fork the original process and continue running in the child (for process group reasons, IIUC), which means that the output stream that was opened in the original process looks open in the child (the shared_ptr is not null, the goodbit is set on the stream), but really it's not open. In that case the behaviour we get is that the first attemp to write to the log file fails completely (no message is written) and we now detect the failure and re-open the log. To the user or admin, it appears as though the output stream never closed.